File: 0663-equal-tree-partition/0663-equal-tree-partition.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> subtreeSum;
    int getTotalSum(TreeNode* node) {
        if(!node) return 0;
        int totalSum = node->val + getTotalSum(node->left) + getTotalSum(node->right);
        subtreeSum.push_back(totalSum);
        return totalSum;
    }
    bool checkEqualTree(TreeNode* root) {
        int totalSum = getTotalSum(root);
        if(totalSum%2!=0)return false;
        subtreeSum.pop_back();
        auto it = find (subtreeSum.begin(), subtreeSum.end(), totalSum/2);
        return it!=subtreeSum.end();
    }
};

File: 0663-equal-tree-partition/0663-equal-tree-partition.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<Integer> subtreeSum;
    public int getTotal(TreeNode node) {
        if(node==null) return 0;
        int totalSum = node.val + getTotal(node.left) + getTotal(node.right);
        subtreeSum.add(totalSum);
        return totalSum;
    }

    public boolean checkEqualTree(TreeNode root) {
        subtreeSum = new ArrayList<>();
        int totalSum = getTotal(root);
        if(totalSum%2!=0) return false; // cannot be divided into equal 2 parts of Sum.
        //remove root
        subtreeSum.remove(subtreeSum.size()-1);
        return subtreeSum.contains(totalSum/2);
    }
}